---
title: "Genome-wide differential gene expression analysis on activated samples of persistent vs resolved subjects at follow-up timepoint"
output: html_notebook
---


```{r}
# Install packages and load libraries
# install.packages("htmltools")
library(htmltools)

# if (!require("BiocManager", quietly = TRUE))
#      install.packages("BiocManager")
# 
#  BiocManager::install("DESeq2")

library( "DESeq2" )
library(ggplot2)

library(stringr) # added by Myra 5/3
```

```{r}
# Read in raw counts matrix
# library(readr)
# df <- read_tsv("C:\Users\Amanda Chan\Documents\440 Project\Dataset\GSE114065_raw_counts_GRCh38.p13_NCBI.tsv.gz")

# Unzips gz file and writes over with unzipped version
# R.utils::gunzip("/Users/Amanda Chan/Documents/440 Project/Dataset/GSE114065_raw_counts_GRCh38.p13_NCBI.tsv.gz")
```

```{r}


#countData <- read.table(file = '/Users/Amanda Chan/Documents/440 Project/Dataset/GSE114065_raw_counts_GRCh38.p13_NCBI.tsv', sep = '\t', header = TRUE)
countData <- read.table(file = '..\\..\\data\\raw\\GSE114065_raw_counts_GRCh38.p13_NCBI.tsv', sep = '\t', header = TRUE) # file path: ..\\..\\ goes up two folders 
```


```{r}
head(countData)

```

```{r}
# Unzip Human Gene Annotation Table gz file and write over with unzipped version
# R.utils::gunzip("/Users/Amanda Chan/Documents/440 Project/Dataset/Human.GRCh38.p13.annot.tsv.gz")
```

```{r}
# Unzip metadata gz file and write over with unzipped version
# R.utils::gunzip("/Users/Amanda Chan/Documents/440 Project/Dataset/GSE114065_series_matrix.txt.gz")
```


```{r}

metaData <- read.csv('..\\..\\data\\raw\\GSE114065_series_matrix.csv', header = TRUE, sep = ",")

# countData <- read.table(file = '..\\..\\data\\raw\\GSE114065_raw_counts_GRCh38.p13_NCBI.tsv', sep = '\t', header = TRUE) # file path: ..\\..\\ goes up two folders

metaData
```

```{r}
# Select Sample_geo_accession (1), Sample_characteristics_ch1_age_yrs (10), Sample_characteristics_ch1_allergy_status (13), and Sample_characteristics_ch1_activation_status (14) rows
coldata <- metaData[c("1", "10", "13", "14"),]
coldata

```

```{r}
coldata_transpose <- t(coldata)

```

```{r}
# Remove rows 1 and 2
coldata_transpose_2 <- coldata_transpose[-c(1, 2),]

# Rename columns
colnames(coldata_transpose_2) <- c("geo_accession", "age", "allergy_status","activation_status")
```

```{r}
# Rename rows to GEO accession names
rownames(coldata_transpose_2) <- coldata_transpose_2[,1]

# Delete column 1 since GEO accession names are now row names
coldata_transpose_3 <- coldata_transpose_2[,-1]
```


PERSISTENT VS RESOLVED ALLERGY AT FOLLOW-UP TIMEPOINT
```{r}

# coldata_transpose_3

# Load dplyr package
library(dplyr)

# Create new metadata table with only samples from the follow-up timepoint (age 2 or 4)
# follow_up_coldata <- filter(coldata_transpose_3, age == '2')
```

```{r}
test_coldata_df <- data.frame(
  age = as.vector(coldata_transpose_3[,1]),
  allergy_status = as.vector(coldata_transpose_3[,2]),
  activation_status = as.vector(coldata_transpose_3[,3]),
  row.names = row.names(coldata_transpose_3)
)

test_coldata_df
```

```{r}
# Create new metadata table with only samples from the follow-up timepoint (age 2 or 4)
follow_up_coldata <- filter(test_coldata_df, age %in% c("2", "4"))
# filter(df, state %in% c("CA", "NY",'DE'))

# Of the follow-up timepoint data, separate into activated and unactivated
follow_up_act_coldata <- filter(follow_up_coldata, activation_status == "1")


# Get sample names of samples at the follow-up timepoint
# sample_label_follow_up <- row.names(follow_up_coldata)
sample_label_follow_up_act <- row.names(follow_up_act_coldata)

```


```{r}
# Create new dataframe with only data samples from the follow-up timepoint (age 2 or 4)
# follow_up_countData <- countData[,c("GeneID", sample_label_follow_up)]

follow_up_act_countData <- countData[,c("GeneID", sample_label_follow_up_act)]



```



```{r}
dds <- DESeqDataSetFromMatrix(countData=follow_up_act_countData, 
                              colData=follow_up_act_coldata, 
                              design=~allergy_status, tidy = TRUE)
dds
```

```{r}
dds <- DESeq(dds)
```

```{r}
# Look at results table
res <- results(dds, alpha=0.05)
head(results(dds, tidy=TRUE))
```

```{r}
# Summary of results
summary(res) #summary of results

```
```{r}
# Sort summary list by p-value
res <- res[order(res$padj),]
head(res)
```

```{r}
# How many adjusted p-values are less than 0.05
sum(res$padj < 0.05, na.rm=TRUE)

```
```{r}
# How many p-values are less than 0.05
sum(res$pvalue < 0.05, na.rm=TRUE)

```
```{r}
# Save results to csv
write.csv(as.data.frame(res), 
         file="..\\..\\data\\results\\differential_gene_expression\\activ_pers_vs_resolved_followup_genomewide.csv")
```



```{r}
# Use plotCounts to compare the normalized counts
# between persistent and resolved groups for top 6 genes
par(mfrow=c(2,3))

plotCounts(dds, gene="7476", intgroup="allergy_status", main = "WNT7A")
plotCounts(dds, gene="10235", intgroup="allergy_status", main = "RASGRP2")
plotCounts(dds, gene="105374981", intgroup="allergy_status", main = "LOC105374981")
plotCounts(dds, gene="348", intgroup="allergy_status", main = "APOE")
plotCounts(dds, gene="255928", intgroup="allergy_status", main = "SYT14")
plotCounts(dds, gene="222545", intgroup="allergy_status", main = "GPRC6A")

```
Volcano Plot - adjusted pvalues
```{r fig.width=7, fig.height=10}

# Save as pdf
pdf(file = "..\\..\\data\\results\\differential_gene_expression\\activ_pers_vs_resolved_followup_genomewide_volc_plot.pdf", width = 7, height = 10)

# Make volcano plot
with(res, plot(log2FoldChange, -log10(padj), pch=20, main="Differential gene expression in future allergy persistent \n vs allergy resolved infants at follow-up timepoint"))

# Add colored points: blue if padj<0.05, red if log2FC>2 and padj<0.05)
with(subset(res, padj<.1 & abs(log2FoldChange)>1), points(log2FoldChange, -log10(padj), pch=20, col="blue"))
with(subset(res, padj<.05 & abs(log2FoldChange)>1), points(log2FoldChange, -log10(padj), pch=20, col="red"))


# Create labels for points
points_label <- subset(res, padj<.1 & abs(log2FoldChange)>1)

# Labels for the significant genes
labels <- c("WNT7A", "RASGRP2", "LOC105374981", "APOE", "SYT14", "GPRC6A")

# Label the significant genes on the volcano plot
with(points_label[1,], text(log2FoldChange, -log10(padj), labels[1], pos=4))
with(points_label[2,], text(log2FoldChange, -log10(padj), labels[2], pos=3))
with(points_label[3,], text(log2FoldChange, -log10(padj), labels[3], pos=4)) 
with(points_label[4,], text(log2FoldChange, -log10(padj), labels[4], pos=2)) 
with(points_label[5:6,], text(log2FoldChange, -log10(padj), labels[5:6], pos=3))
# pos specifies the position of the labels relative to the points


# Add legend
legend("topright", legend = c("padj < 0.05 and abs(log2FoldChange) > 1", "padj < 0.1 and abs(log2FoldChange) > 1"), col = c("red", "blue"), cex=0.8, pch=20)


dev.off()

```




# added by Myra 5/3
# https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/05_Annotation_and_Visualisation.nb.html

```{r}
library(biomaRt)
ensembltest = useMart("ENSEMBL_MART_ENSEMBL")
listDatasets(ensembltest) %>% 
    filter(str_detect(description, "Human"))
ensembl = useDataset("hsapiens_gene_ensembl", mart=ensembltest)
```

```{r}

res = read.csv("..\\..\\data\\results\\differential_gene_expression\\activ_pers_vs_resolved_followup_genomewide.csv", row.names=1)

listFilters(ensembl) %>% 
    filter(str_detect(name, "ensembl"))

# Set the filter type and values
filterType = "entrezgene_id"
filterValues = rownames(res)
print(filterType)
# check the available "attributes" - things you can retreive
listAttributes(ensembl) %>% 
    head(100)

#sink(file = "ensembl_db_attributes.txt")
lapply(listAttributes(ensembl), write, "ensembl_db_attributes.txt", append=TRUE, ncolumns=1)
```


```{r}
# Set the list of attributes
attributeNames <- c('ensembl_gene_id', 'entrezgene_id', 'external_gene_name', 'ensembl_transcript_id') 
# run the query
annot <- getBM(attributes=attributeNames, 
               filters = filterType, 
               values = filterValues, 
               mart = ensembl)

head(annot)

```

```{r}
# Save results to csv
write.csv(as.data.frame(annot), 
         file="..\\..\\data\\results\\differential_gene_expression\\biomart_geneid_lookuptable.csv")
```